src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -314,17 +345,123 @@ public TypeOverloadInfo Aggregate(INamedTypeSymbol type)
+
+                public ReferenceCacheEntry ToReferenceCacheEntry()
+                {
+                    return new ReferenceCacheEntry(
+                        _checksum,
+                        _isCSharp,
+                        _commonItemsBuilder.ToImmutableAndFree(),
+                        _genericItemsBuilder.ToImmutableAndFree(),
+                        _attributeWithSuffixItemsBuilder.ToImmutableAndFree());
+                }
+
+                public void AddItem(INamedTypeSymbol symbol, string containingNamespace, bool isPublic)
+                {
+                    ArrayBuilder<TypeImportCompletionItemInfo> correspondingBuilder;
+
+                    // Attribute type can't be generic
+                    if (symbol.Arity > 0)
+                    {
+                        correspondingBuilder = _genericItemsBuilder;
+                    }
+                    else if (IsAttributeWithAttributeSuffix(symbol))
+                    {
+                        correspondingBuilder = _attributeWithSuffixItemsBuilder;
+                    }
+                    else
+                    {
+                        correspondingBuilder = _commonItemsBuilder;
+                    }
+
+                    var item = TypeImportCompletionItem.Create(symbol, containingNamespace, _isCSharp);
+                    item.IsCached = true;
+                    correspondingBuilder.Add(new TypeImportCompletionItemInfo(item, isPublic));
+
+                    static bool IsAttributeWithAttributeSuffix(INamedTypeSymbol symbol)
+                    {
+                        // Do the symbol textual check first. Then the more expensive symbolic check.
+                        return symbol.Name.HasAttributeSuffix(isCaseSensitive: false) && symbol.IsAttribute();
+                    }
+                }
+            }
+
+            private ReferenceCacheEntry(
                 Checksum checksum,
-                ImmutableArray<TypeImportCompletionItemInfo> cachedItems)
+                bool isCSharp,
+                ImmutableArray<TypeImportCompletionItemInfo> commonItems,
+                ImmutableArray<TypeImportCompletionItemInfo> genericItems,
+                ImmutableArray<TypeImportCompletionItemInfo> attributeItems)
             {
                 Checksum = checksum;
-                CachedItems = cachedItems;
+                IsCSharp = isCSharp;
+
+                CommonItems = commonItems;
+                AttributeItems = attributeItems;
+                GenericItems = genericItems;
             }
 
+            public bool IsCSharp { get; }
+
             public Checksum Checksum { get; }
 
-            public ImmutableArray<TypeImportCompletionItemInfo> CachedItems { get; }
+            public ImmutableArray<TypeImportCompletionItemInfo> CommonItems { get; }
+
+            private ImmutableArray<TypeImportCompletionItemInfo> GenericItems { get; }
+
+            // Attribute types can't be generic
```

Comment:
note: it's quite likely this will change in C# 8 timeframe.  there are open PRs/designs for making this possible in the language

Reply:
@jcouv Is the generic attribute change targeting 16.3?

Reply:
No, it will come after 16.3.
The [PR](https://github.com/dotnet/roslyn/pull/26337) for that feature is targeting a feature branch.

Reply:
Thanks. I will file an issue to track this.

Reply:
@jcouv Could you please take a look at the compiler change? It's just adding a string helper. Thanks!